gector/gec_predictor.py

```diff
@@ -143,15 +157,102 @@ def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
                 - corrected_words
             For an explanation of each of these see `Seq2Labels.decode()`.
         """
-        for i in range(self._iterations):
-            outputs = self._model.forward_on_instances(instances)
-            corrected_instances = []
-            for output in outputs:
-                tokens = [Token(word) for word in output["corrected_words"]]
-                instance = self._dataset_reader.text_to_instance(tokens)
-                corrected_instances.append(instance)
-            instances = corrected_instances
-        return sanitize(outputs)
+
+        # Make deep copy of batch
+        final_batch = instances[:]
+
+        # Create list to store final predictions
+        final_outputs = [None] * len(instances)
+
+        # This dictionary keeps track of predictions made in every iteration
+        prev_preds_dict = {}
+
+        # This list contains IDs of sentences to be passed into model
+        pred_ids = []
+
+        # Populating `prev_preds_dict` and `pred_ids`
+        for id, instance in enumerate(final_batch):
+            prev_preds_dict[id] = [instance.fields["tokens"].tokens]
+            # If len(tokens) is less than 4 ($START + 3 tokens)
+            # we will not correct it.
+            # It is directly written to output.
+            if len(instance.fields["tokens"].tokens) < 4:
+                final_outputs[id] = {
+                    "logits_labels": None,
+                    "logits_d_tags": None,
+                    "class_probabilities_labels": None,
+                    "class_probabilities_d_tags": None,
+                    "max_error_probability": None,
+                    "words": instance.fields["tokens"].tokens[1:],
+                    "labels": None,
+                    "d_tags": None,
+                    "corrected_words": instance.fields["tokens"].tokens[1:],
+                }
+            else:
+                pred_ids.append(id)
+
+        # Applying correction model multiple times
+        for _ in range(self._iterations):
+
+            # If no sentences need to be passed into model
+            if len(pred_ids) == 0:
+                break
+
+            # Create batch of instances to be passed into model
+            orig_batch = [final_batch[pred_id] for pred_id in pred_ids]
+
+            # Pass into model
+            outputs = self._model.forward_on_instances(orig_batch)
+
+            new_pred_ids = []
+
+            # Output_ID and Pred_ID in pred_ids
+            for op_ind, pred_id in enumerate(pred_ids):
+
+                # Update final outputs
+                final_outputs[pred_id] = outputs[op_ind]
+                orig = final_batch[pred_id]
+
+                # Create tokens from corrected words for next iter
+                tokens = [
+                    Token(word)
+                    for word in ["$START"] + outputs[op_ind]["corrected_words"]
```

Comment:
If I was betting, I'd say that this is where the discrepancies were coming from.

Reply:
Yeah, it was a combination of things, a lot of it to do with indices and START tokens etc.